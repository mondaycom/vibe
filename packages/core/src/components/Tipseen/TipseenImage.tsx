import cx from "classnames";
import React from "react";
import styles from "./TipseenImage.module.scss";
import TipseenMedia from "./TipseenMedia/TipseenMedia";

export interface TipseenImageProps {
  /**
   * The source URL of the image.
   */
  src?: string | undefined;
  /**
   * The alternative text for the image, used for accessibility.
   */
  alt?: string;
  /**
   * Class name applied to the image element.
   */
  className?: string;
  /**
   * Class name applied to the TipseenMedia container.
   */
  tipseenMediaClassName?: string;
}

const TipseenImage: React.FC<TipseenImageProps> = ({ src, alt, className, tipseenMediaClassName }) => {
  return (
    // TODO future breaking change where we remove the TipseenMedia component and we remove TipseenImage?
    <TipseenMedia className={tipseenMediaClassName}>
      <img src={src} alt={alt} className={cx(styles.tipseenImage, className)} />
    </TipseenMedia>
  );
};

export default TipseenImage;
