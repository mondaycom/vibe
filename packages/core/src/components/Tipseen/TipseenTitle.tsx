import React, { FC } from "react";
import { VibeComponentProps } from "../../types";
import { getTestId } from "../../tests/test-ids-utils";
import { ComponentDefaultTestId } from "../../tests/constants";
import Text from "../Text/Text";
import cx from "classnames";
import styles from "./TipseenTitle.module.scss";

export interface TipseenTitleProps extends VibeComponentProps {
  /**
   * The title text displayed in the Tipseen.
   */
  text?: string;
}

const TipseenTitle: FC<TipseenTitleProps> = ({ text, className, id, "data-testid": dataTestId }) => {
  return text ? (
    <Text
      type="text1"
      weight="bold"
      role="heading"
      color="inherit"
      aria-level={3}
      maxLines={2}
      className={cx(styles.title, className)}
      id={id}
      data-testid={dataTestId || getTestId(ComponentDefaultTestId.TIPSEEN_TITLE, id)}
    >
      {text}
    </Text>
  ) : null;
};

export default TipseenTitle;
