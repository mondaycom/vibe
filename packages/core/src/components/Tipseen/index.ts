export { default as Tipseen, type TipseenProps } from "./Tipseen";
export { default as TipseenContent, type TipseenContentProps } from "./TipseenContent";
export { default as TipseenWizard, type TipseenWizardProps } from "./TipseenWizard";
export { default as TipseenImage, type TipseenImageProps } from "./TipseenImage";
export { default as TipseenMedia, type TipseenMediaProps } from "./TipseenMedia/TipseenMedia";

export * from "./Tipseen.types";
