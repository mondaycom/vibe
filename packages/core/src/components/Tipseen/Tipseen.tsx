import { forwardRef, Fragment, ReactElement, useEffect, useMemo, useRef, useState } from "react";
import cx from "classnames";
import { AnimationType as AnimationTypeEnum, HideShowEvent as HideShowEventEnum } from "../Dialog/DialogConstants";
import { DialogAnimationType, DialogTriggerEvent } from "../Dialog/Dialog.types";
import useMergeRef from "../../hooks/useMergeRef";
import Tooltip from "../../components/Tooltip/Tooltip";
import IconButton from "../../components/IconButton/IconButton";
import { CloseSmall } from "@vibe/icons";
import TipseenTitle from "./TipseenTitle";
import {
  TIPSEEN_CLOSE_BUTTON_ARIA_LABEL,
  TipseenCloseButtonTheme as TipseenCloseButtonThemeEnum,
  TipseenColor as TipseenColorEnum
} from "./TipseenConstants";
import { TipseenCloseButtonTheme, TipseenColor } from "./Tipseen.types";
import { ElementContent, VibeComponentProps, withStaticProps } from "../../types";
import { MoveBy } from "../../types/MoveBy";
import { Modifier } from "react-popper";
import { ComponentDefaultTestId } from "../../tests/constants";
import { getTestId } from "../../tests/test-ids-utils";
import Text from "../Text/Text";
import styles from "./Tipseen.module.scss";
import { ButtonColor } from "../Button/ButtonConstants";
import React from "react";
import { TooltipPositions } from "../Tooltip/Tooltip.types";
import { TooltipPositions as TooltipPositionsEnum } from "../Tooltip/TooltipConstants";

export interface TipseenProps extends VibeComponentProps {
  /**
   * Class name applied to the Tipseen title.
   */
  titleClassName?: string;
  /**
   * The position of the Tipseen relative to the target element.
   */
  position?: TooltipPositions;
  /**
   * The animation type used for showing/hiding the Tipseen.
   */
  animationType?: DialogAnimationType;
  /**
   * The delay in milliseconds before hiding the Tipseen.
   */
  hideDelay?: number;
  /**
   * The delay in milliseconds before showing the Tipseen.
   */
  showDelay?: number;
  /**
   * The title text of the Tipseen.
   */
  title?: string;
  /**
   * If true, hides the close button.
   */
  hideCloseButton?: boolean;
  /**
   * The child element that triggers the Tipseen.
   */
  children?: ReactElement;
  /**
   * The CSS selector of the container where the Tipseen should be rendered.
   */
  containerSelector?: string;
  /**
   * Events that trigger hiding the Tipseen.
   */
  hideTrigger?: DialogTriggerEvent | Array<DialogTriggerEvent>;
  /**
   * Events that trigger showing the Tipseen.
   */
  showTrigger?: DialogTriggerEvent | Array<DialogTriggerEvent>;
  /**
   * The width of the Tipseen.
   */
  width?: number;
  /**
   * Offset values for positioning adjustments.
   */
  moveBy?: MoveBy;
  /**
   * If true, hides the Tipseen when the reference element is hidden.
   */
  hideWhenReferenceHidden?: boolean;
  /**
   * Class name applied to the reference wrapper element.
   */
  referenceWrapperClassName?: string;
  /**
   * If false, hides the arrow of the Tipseen.
   */
  tip?: boolean;
  /**
   * Class name applied to the Tipseen arrow.
   */
  tooltipArrowClassName?: string;
  /**
   * Custom Popper.js modifiers.
   * https://popper.js.org/docs/v2/modifiers/
   */
  modifiers?: Array<Modifier<unknown>>;
  /**
   * The aria-label for the close button.
   */
  closeAriaLabel?: string;
  /**
   * Callback fired when the Tipseen is closed.
   */
  onClose?: (event?: React.MouseEvent<HTMLButtonElement>) => void;
  /**
   * The content displayed inside the Tipseen.
   */
  content: ElementContent;
  /**
   * The theme of the Tipseen close button.
   */
  closeButtonTheme?: TipseenCloseButtonTheme;
  /**
   * If true, renders the Tipseen as a floating element without a reference.
   */
  floating?: boolean;
  /**
   * The color theme of the Tipseen.
   */
  color?: TipseenColor;
}

export const TipseenContext = React.createContext<TipseenColor>("primary");

const Tipseen = forwardRef(
  (
    {
      className,
      id,
      position = "bottom",
      animationType = "expand",
      hideDelay = 0,
      showDelay = 100,
      title,
      titleClassName,
      hideCloseButton,
      closeButtonTheme = "light",
      onClose,
      closeAriaLabel,
      children = null,
      content,
      containerSelector,
      hideTrigger = [],
      showTrigger = [],
      width,
      moveBy,
      hideWhenReferenceHidden = false,
      referenceWrapperClassName,
      tip = true,
      tooltipArrowClassName,
      modifiers = [],
      floating = false,
      color: colorProp,
      "data-testid": dataTestId
    }: TipseenProps,
    ref: React.ForwardedRef<HTMLElement>
  ) => {
    const color = colorProp ?? "inverted";

    const defaultDelayOpen =
      Array.isArray(showTrigger) && Array.isArray(hideTrigger) && showTrigger.length === 0 && showDelay > 0;

    const componentRef = useRef(null);
    const mergedRef = useMergeRef(ref, componentRef);
    const [delayedOpen, setDelayOpen] = useState(!defaultDelayOpen);
    const overrideCloseAriaLabel = closeAriaLabel || TIPSEEN_CLOSE_BUTTON_ARIA_LABEL;

    useEffect(() => {
      let timeout: NodeJS.Timeout;
      if (showDelay) {
        timeout = setTimeout(() => {
          setDelayOpen(true);
        }, showDelay);
      }
      return () => {
        clearTimeout(timeout);
      };
    }, [showDelay, setDelayOpen]);

    const textColor = useMemo(() => {
      return color === "inverted" ? "onInverted" : "onPrimary";
    }, [color]);
    const closeButtonColor = useMemo(() => {
      if (closeButtonTheme === "light") {
        return color === "inverted" ? ButtonColor.ON_INVERTED_BACKGROUND : ButtonColor.ON_PRIMARY_COLOR;
      } else {
        return closeButtonTheme;
      }
    }, [color, closeButtonTheme]);

    const TipseenWrapper = ref || id ? "div" : Fragment;
    const tooltipContent = (
      <div>
        <div className={cx(styles.tipseenHeader)}>
          {hideCloseButton ? null : (
            <IconButton
              hideTooltip
              className={cx(styles.tipseenCloseButton, {
                [styles.dark]: closeButtonTheme === "dark" || closeButtonTheme === "fixed-dark"
              })}
              onClick={onClose}
              size="xs"
              kind="tertiary"
              // @ts-ignore
              color={closeButtonColor}
              ariaLabel={overrideCloseAriaLabel}
              icon={CloseSmall}
            />
          )}
          <TipseenTitle text={title} className={cx(styles.tipseenTitle, titleClassName)} />
        </div>
        <Text color={textColor} type="text2" element="p" className={cx(styles.tipseenContent)}>
          <TipseenContext.Provider value={color}>{content}</TipseenContext.Provider>
        </Text>
      </div>
    );

    return (
      <TipseenWrapper ref={mergedRef} id={id} data-testid={dataTestId || getTestId(ComponentDefaultTestId.TIPSEEN, id)}>
        <Tooltip
          className={cx(styles.tipseenWrapper, className, {
            [styles.tipseenWrapperWithoutCustomWidth]: !width,
            [styles.floating]: floating
          })}
          maxWidth={width}
          arrowClassName={tooltipArrowClassName}
          style={width ? { width } : undefined}
          shouldShowOnMount={!defaultDelayOpen}
          position={position}
          animationType={animationType}
          hideDelay={hideDelay}
          showDelay={0}
          hideTrigger={hideTrigger}
          showTrigger={showTrigger}
          showOnDialogEnter={false}
          content={tooltipContent}
          theme={color === "inverted" ? "dark" : "primary"}
          containerSelector={containerSelector}
          disableDialogSlide={false}
          moveBy={moveBy}
          hideWhenReferenceHidden={hideWhenReferenceHidden}
          referenceWrapperClassName={referenceWrapperClassName}
          tip={tip && !floating}
          modifiers={modifiers}
          open={defaultDelayOpen ? delayedOpen : undefined}
          forceRenderWithoutChildren={floating}
        >
          {children}
        </Tooltip>
      </TipseenWrapper>
    );
  }
);

interface TipseenStaticProps {
  closeButtonThemes: typeof TipseenCloseButtonThemeEnum;
  animationTypes: typeof AnimationTypeEnum;
  hideShowTriggers: typeof HideShowEventEnum;
  colors: typeof TipseenColorEnum;
  positions: typeof TooltipPositionsEnum;
}

export default withStaticProps<TipseenProps, TipseenStaticProps>(Tipseen, {
  closeButtonThemes: TipseenCloseButtonThemeEnum,
  animationTypes: AnimationTypeEnum,
  hideShowTriggers: HideShowEventEnum,
  colors: TipseenColorEnum,
  positions: TooltipPositionsEnum
});
