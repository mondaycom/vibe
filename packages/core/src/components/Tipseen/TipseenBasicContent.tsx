import React from "react";
import cx from "classnames";
import TipseenTitle from "./TipseenTitle";
import { VibeComponentProps } from "../../types";
import { ElementContent } from "../../types/ElementContent";
import { ComponentDefaultTestId } from "../../tests/constants";
import { getTestId } from "../../tests/test-ids-utils";
import styles from "./TipseenBasicContent.module.scss";

export interface TipseenBasicContentProps extends VibeComponentProps {
  /**
   * The title text displayed in the Tipseen content.
   */
  title?: string;
  /**
   * Class name applied to the title.
   */
  titleClassName?: string;
  /**
   * The content inside the Tipseen.
   */
  children?: ElementContent | ElementContent[];
}

const TipseenBasicContent: React.FC<TipseenBasicContentProps> = ({
  title,
  children = null,
  titleClassName,
  className,
  id,
  "data-testid": dataTestId
}) => {
  return (
    <div
      className={cx(styles.tipseenBasicContent, className)}
      id={id}
      data-testid={dataTestId || getTestId(ComponentDefaultTestId.TIPSEEN_CONTENT, id)}
    >
      <TipseenTitle text={title} className={titleClassName} />
      {children}
    </div>
  );
};

export default TipseenBasicContent;
