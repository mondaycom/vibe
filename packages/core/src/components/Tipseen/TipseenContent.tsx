import React, { FC, useContext, useMemo } from "react";
import cx from "classnames";
import Button from "../../components/Button/Button";
import { DISMISS_BUTTON_TEXT, SUBMIT_BUTTON_TEXT, TipseenColor } from "./TipseenConstants";
import TipseenBasicContent from "./TipseenBasicContent";
import { VibeComponentProps } from "../../types";
import { ElementContent } from "../../types/ElementContent";
import styles from "./TipseenContent.module.scss";
import { getTestId } from "../../tests/test-ids-utils";
import { ComponentDefaultTestId } from "../../tests/constants";
import { TipseenContext } from "./Tipseen";
import { ButtonColor } from "../Button/ButtonConstants";
import { SubIcon } from "../../types";

export interface TipseenContentProps extends VibeComponentProps {
  /**
   * The title text displayed in the Tipseen content.
   */
  title?: string;
  /**
   * Class name applied to the Tipseen title.
   */
  titleClassName?: string;
  /**
   * If true, hides the dismiss button.
   */
  hideDismiss?: boolean;
  /**
   * The content inside the Tipseen.
   */
  children?: ElementContent;
  /**
   * If true, hides the submit button.
   */
  hideSubmit?: boolean;
  /**
   * The text displayed on the submit button.
   */
  submitButtonText?: string;
  /**
   * Icon to display in the submit button
   */
  submitButtonIcon?: SubIcon;
  /**
   * Callback fired when the submit button is clicked.
   */
  onSubmit?: (event: React.MouseEvent) => void;
  /**
   * The text displayed on the dismiss button.
   */
  dismissButtonText?: string;
  /**
   * Callback fired when the dismiss button is clicked.
   */
  onDismiss?: (event: React.MouseEvent) => void;
}

const TipseenContent: FC<TipseenContentProps> = ({
  id,
  title,
  titleClassName,
  children = null,
  hideDismiss = true,
  hideSubmit,
  submitButtonText = SUBMIT_BUTTON_TEXT,
  submitButtonIcon,
  onSubmit,
  dismissButtonText = DISMISS_BUTTON_TEXT,
  onDismiss
}) => {
  const color = useContext(TipseenContext);
  const buttonColor = useMemo(() => {
    return color === TipseenColor.INVERTED ? ButtonColor.ON_INVERTED_BACKGROUND : ButtonColor.ON_PRIMARY_COLOR;
  }, [color]);

  return (
    <TipseenBasicContent title={title} titleClassName={titleClassName} id={id}>
      {children ? <span>{children}</span> : null}
      <div className={cx(styles.buttons)}>
        {hideDismiss ? null : (
          <Button
            kind="tertiary"
            color={buttonColor}
            className={styles.dismiss}
            size="small"
            onClick={onDismiss}
            data-testid={getTestId(ComponentDefaultTestId.TIPSEEN_CONTENT_DISMISS)}
          >
            {dismissButtonText}
          </Button>
        )}
        {hideSubmit ? null : (
          <Button
            kind="primary"
            color={buttonColor}
            size="small"
            onClick={onSubmit}
            data-testid={getTestId(ComponentDefaultTestId.TIPSEEN_CONTENT_SUBMIT)}
            leftIcon={submitButtonIcon}
          >
            {submitButtonText}
          </Button>
        )}
      </div>
    </TipseenBasicContent>
  );
};

export default TipseenContent;
