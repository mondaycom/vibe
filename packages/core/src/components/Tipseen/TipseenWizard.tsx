import React, { FC, useContext, useMemo } from "react";
import cx from "classnames";
import Steps, { StepsProps } from "../Steps/Steps";
import TipseenBasicContent from "./TipseenBasicContent";
import styles from "./TipseenWizard.module.scss";
import { TipseenContext } from "./Tipseen";
import { TipseenColor } from "./TipseenConstants";
import { ButtonSize, ButtonType } from "../Button";

const FINISH_TEXT = "Got it";

export interface TipseenWizardProps extends StepsProps {
  /**
   * The title text displayed in the Tipseen Wizard.
   */
  title?: string;
  /**
   * Class name applied to the Tipseen title.
   */
  titleClassName?: string;
  /**
   * Callback fired when the wizard is completed.
   */
  onFinish?: (e: React.MouseEvent | React.KeyboardEvent) => void;
}

const TipseenWizard: FC<TipseenWizardProps> = ({ id, title, onFinish, titleClassName, className, ...stepsProps }) => {
  const overrideStepsProps = stepsProps as StepsProps;
  const color = useContext(TipseenContext);
  const buttonColor = useMemo(() => {
    return color === TipseenColor.INVERTED ? "on-inverted-background" : "on-primary-color";
  }, [color]);

  const nextButtonProps = useMemo(
    () => ({
      kind: "primary" as ButtonType,
      size: "small" as ButtonSize
    }),
    []
  );
  const backButtonProps = useMemo(
    () => ({
      size: "small" as ButtonSize
    }),
    []
  );
  const finishButtonProps = useMemo(
    () => ({
      kind: "primary" as ButtonType,
      size: "small" as ButtonSize,
      children: FINISH_TEXT
    }),
    []
  );
  return (
    <TipseenBasicContent
      title={title}
      className={cx(styles.tipseenWizard, className)}
      titleClassName={titleClassName}
      id={id || "wizard"}
    >
      <Steps
        className={cx(styles.tipseenWizardWizard)}
        color={buttonColor}
        isContentOnTop
        areButtonsIconsHidden
        backButtonProps={backButtonProps}
        nextButtonProps={nextButtonProps}
        finishButtonProps={finishButtonProps}
        onFinish={onFinish}
        {...overrideStepsProps}
      />
    </TipseenBasicContent>
  );
};
export default TipseenWizard;
