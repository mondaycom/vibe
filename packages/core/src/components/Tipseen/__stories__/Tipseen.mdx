import { Meta } from "@storybook/blocks";
import { Link } from "vibe-storybook-components";
import Tipseen from "../Tipseen";
import TipseenContent from "../TipseenContent";
import TipseenWizard from "../TipseenWizard";
import Icon from "../../Icon/Icon";
import { Board, BoardPrivate } from "@vibe/icons";
import { modifiers, TipCheckYourself, wizardContent } from "./Tipseen.stories.helpers";
import {
  ATTENTION_BOX,
  CHIP,
  TOOLTIP
} from "../../../storybook/components/related-components/component-description-map";
import * as TipseenStories from "./Tipseen.stories";
import { tipseenTemplate } from "./Tipseen.stories";

<Meta of={TipseenStories} />

# Tipseen

- [Overview](#overview)
- [Props](#props)
- [Usage](#usage)
- [Variants](#variants)
- [Do‚Äôs and don‚Äôts](#dos-and-donts)
- [Use cases and examples](#use-cases-and-examples)
- [Related components](#related-components)
- [Feedback](#feedback)

## Overview

Tipseen is a virtual unboxing experience that helps users get started with the system and discover new features.

<Canvas of={TipseenStories.Overview} />

### Import

```js
import { Tipseen, TipseenMedia, TipseenImage, TipseenContent, TipseenWizard } from "@vibe/core";
```

## Props

<PropsTable />

## Usage

<UsageGuidelines
  guidelines={[
    "Use for onboarding screens, new features discovery, or any guidance a user needs.",
    "Use when the user is not yet familiar with the system‚Äôs UI or ready to learn about it.",
    "The tip will appear until closed by an X button or call to action button."
  ]}
/>

<TipCheckYourself />

## Variants

### Colors

Tipseens come in 2 colors: inverted and primary. \
The inverted type is the default. \
Use the inverted color for feature discovery, or to give the user general guidance. \
Use the primary color to bring attention to updates about your product offering.

<Canvas of={TipseenStories.Colors} />

### Tipseen with a wizard

Use Tipseen with a wizard when you want to teach something in steps.

<Canvas of={TipseenStories.TipseenWithAWizard} />

### Tipseen with image

<Canvas of={TipseenStories.TipseenWithImage} />

### Tipseen with custom media

Wrap your custom media with `TipseenMedia` component. This use case is good for when using Lottie animations or other media that is not necessarily an image.

<Canvas of={TipseenStories.TipseenWithCustomMedia} />

### Floating Tipseen

Use Floating Tipseen in case where the guidance is not relevant to a specific UI element, but general.
In this case, the Tipseen position will be the right corner of the screen and without an arrow tip.

<Canvas of={TipseenStories.FloatingTipseen} story={{ height: "500px" }} />

## Do‚Äôs and Don‚Äôts

<ComponentRules
  className="monday-storybook-tipseen_rules"
  rules={[
    {
      positive: {
        component: (
          <div className="monday-storybook-tipseen_column-wrapper right">
            <Tipseen
              modifiers={modifiers}
              position="right"
              content={
                <TipseenWizard title="This is a title" steps={wizardContent} activeStepIndex={4} onFinish={() => {}} />
              }
            >
              <div className="monday-storybook-tipseen_inline-container right" />{" "}
            </Tipseen>
          </div>
        ),
        description:
          "When using a Tipseen wizard, allow the user to complete the process and close the Tipseen by leaving the last CTA enabled."
      },
      negative: {
        component: (
          <div className="monday-storybook-tipseen_column-wrapper right">
            <Tipseen
              modifiers={modifiers}
              position="right"
              content={<TipseenWizard title="This is a title" steps={wizardContent} activeStepIndex={4} />}
            >
              <div className="monday-storybook-tipseen_inline-container right" />{" "}
            </Tipseen>
          </div>
        ),
        description: "Don‚Äôt finish the Tipseen wizard process with a disabled CTA."
      }
    },
    {
      positive: {
        component: (
          <div className="monday-storybook-tipseen_box">
            <div className="monday-storybook-tipseen_column-item">
              <Icon icon={Board} iconSize="24" />
              <span>üí´ Research tasks</span>
            </div>
            <Tipseen
              modifiers={modifiers}
              position="right"
              content={<TipseenContent title="Step 1 of 5">Click on the ‚ÄúSales‚Äù board to get started.</TipseenContent>}
            >
              <div className="monday-storybook-tipseen_column-item monday-storybook-tipseen_hovered">
                <Icon icon={Board} iconSize="24" />
                <span>Sales Pipeline</span>
              </div>
            </Tipseen>
            <div className="monday-storybook-tipseen_column-item">
              <Icon icon={BoardPrivate} iconSize="24" />
              <span>Marketing</span>
            </div>
          </div>
        ),
        description:
          "Use tipseen in order to guide the user through a new feature or a place they‚Äôre not familiar with in the system."
      },
      negative: {
        component: (
          <div className="monday-storybook-tipseen_column-wrapper">
            <Tipseen
              modifiers={modifiers}
              position="top"
              content={
                <TipseenContent>
                  You still have 14 days on your free trial, enjoy it! At the end you can choose your favorite plan.
                </TipseenContent>
              }
            >
              <div className="monday-storybook-tipseen_inline-container" />
            </Tipseen>
          </div>
        ),
        description: (
          <>
            Don‚Äôt use tipseen to provide additional information on a familiar UI. Use a{" "}
            <StorybookLink page="Components/Tooltip">Tooltip</StorybookLink> or{" "}
            <StorybookLink page="Components/AttentionBox">Attention box</StorybookLink> instead.
          </>
        )
      }
    },
    {
      positive: {
        component: (
          <div className="monday-storybook-tipseen_box">
            <Tipseen
              modifiers={modifiers}
              position="right"
              content={<TipseenContent title="Step 1 of 5">Click on the ‚ÄúSales‚Äù board to get started.</TipseenContent>}
            >
              <div className="monday-storybook-tipseen_column-item monday-storybook-tipseen_hovered">
                <Icon icon={Board} iconSize="24" />
                <span>Sales Pipeline</span>
              </div>
            </Tipseen>
          </div>
        ),
        description: "Add a pointer to indicate the specific component that the tipseen comes from."
      },
      negative: {
        component: (
          <div className="monday-storybook-tipseen_box">
            <Tipseen
              modifiers={modifiers}
              position="right"
              tip={false}
              content={<TipseenContent title="Step 1 of 5">Click on the ‚ÄúSales‚Äù board to get started.</TipseenContent>}
            >
              <div className="monday-storybook-tipseen_column-item monday-storybook-tipseen_hovered">
                <Icon icon={Board} iconSize="24" />
                <span>Sales Pipeline</span>
              </div>
            </Tipseen>
          </div>
        ),
        description: "Don‚Äôt use a tipseen without a pointer, unless it‚Äôs a general tipseen for an entire screen."
      }
    }
  ]}
/>

## Related components

<RelatedComponents componentsNames={[TOOLTIP, ATTENTION_BOX, CHIP]} />
