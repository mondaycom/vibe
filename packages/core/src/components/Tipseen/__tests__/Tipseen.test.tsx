import { vi, describe, it, expect, beforeEach, afterEach } from "vitest";
import React from "react";
import { fireEvent, render, waitFor } from "@testing-library/react";
import { act } from "@testing-library/react-hooks";
import TipseenContent from "../TipseenContent";
import Tipseen from "../Tipseen";
import { DISMISS_BUTTON_TEXT, SUBMIT_BUTTON_TEXT } from "../TipseenConstants";
import renderer from "react-test-renderer";
import TipseenTitle from "../TipseenTitle";

const tipseenMockChildren = <div className="monday-storybook-tipseen_container" />;

describe("Snapshot tests", () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  describe("TipseenTitle", () => {
    it("should render null without text", () => {
      const tree = renderer.create(<TipseenTitle />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("should render correctly with given text", () => {
      const tree = renderer.create(<TipseenTitle text="I'm a title" />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("Tipseen content tests", () => {
    it("renders correctly without props", () => {
      const { asFragment } = render(<TipseenContent />);
      expect(asFragment()).toMatchSnapshot();
    });
    it("renders correctly with dismiss", () => {
      const tree = renderer.create(<TipseenContent hideDismiss={false} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it("renders correctly without submit", () => {
      const tree = renderer.create(<TipseenContent hideSubmit />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("Tipseen tests", () => {
    it("renders correctly without p" + "rops", () => {
      const tree = renderer.create(<Tipseen content="content">{tipseenMockChildren}</Tipseen>).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it("renders correctly without close", () => {
      const { asFragment } = render(
        <Tipseen content="content" showDelay={0} hideCloseButton>
          {tipseenMockChildren}
        </Tipseen>
      );
      expect(asFragment()).toMatchSnapshot();
    });
    it("renders correctly with dark close button theme", async () => {
      const { asFragment } = render(
        <Tipseen content="content" showDelay={0} closeButtonTheme="dark">
          {tipseenMockChildren}
        </Tipseen>
      );
      await waitFor(() => {
        expect(asFragment()).toBeTruthy();
      });
      expect(asFragment()).toMatchSnapshot();
    });
    it("renders correctly with floating variation", () => {
      const { container } = render(<Tipseen content="content" floating />);

      // Run timers to ensure any async operations complete
      act(() => {
        vi.runAllTimers();
      });

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});

describe("Integration Tests", () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  describe("Tipseen tests", () => {
    it("call onClose function when click on close button", () => {
      const onClickMock = vi.fn();
      const { getByLabelText } = render(
        <Tipseen content="content" onClose={onClickMock}>
          <div />
        </Tipseen>
      );

      waitFor(() => {
        fireEvent.click(getByLabelText("Close"));
        expect(onClickMock.mock.calls.length).toBe(1);
      });
    });
  });

  describe("Tipseen content tests", () => {
    it("call onDismiss function when click on dismiss button", () => {
      const onDismissMock = vi.fn();
      const { getByText } = render(
        <TipseenContent hideDismiss={false} onDismiss={onDismissMock}>
          content
        </TipseenContent>
      );
      const dismissButton = getByText(DISMISS_BUTTON_TEXT);

      act(() => {
        fireEvent.click(dismissButton);
      });
      expect(onDismissMock.mock.calls.length).toBe(1);
    });

    it("call onSubmit function when click on dismiss button", () => {
      const onSubmitMock = vi.fn();
      const { getByText } = render(<TipseenContent onSubmit={onSubmitMock}>content</TipseenContent>);
      const submitButton = getByText(SUBMIT_BUTTON_TEXT);

      act(() => {
        fireEvent.click(submitButton);
      });
      expect(onSubmitMock.mock.calls.length).toBe(1);
    });
  });
});
