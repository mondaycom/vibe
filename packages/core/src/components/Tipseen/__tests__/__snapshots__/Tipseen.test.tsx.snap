// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot tests > Tipseen content tests > renders correctly with dismiss 1`] = `
<div
  className="tipseenBasicContent"
  data-testid="tipseen-content"
>
  <div
    className="buttons"
  >
    <button
      aria-busy={false}
      aria-disabled={false}
      className="dismiss button sizeSmall kindTertiary colorOnPrimaryColor"
      data-testid="tipseen-content-dismiss"
      data-vibe="Button"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="button"
    >
      Dismiss
    </button>
    <button
      aria-busy={false}
      aria-disabled={false}
      className="button sizeSmall kindPrimary colorOnPrimaryColor"
      data-testid="tipseen-content-submit"
      data-vibe="Button"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="button"
    >
      Submit
    </button>
  </div>
</div>
`;

exports[`Snapshot tests > Tipseen content tests > renders correctly without props 1`] = `
<DocumentFragment>
  <div
    class="tipseenBasicContent"
    data-testid="tipseen-content"
  >
    <div
      class="buttons"
    >
      <button
        aria-busy="false"
        aria-disabled="false"
        class="button sizeSmall kindPrimary colorOnPrimaryColor"
        data-testid="tipseen-content-submit"
        data-vibe="Button"
        style="color: rgba(0, 0, 0, 0);"
        type="button"
      >
        Submit
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Snapshot tests > Tipseen content tests > renders correctly without submit 1`] = `
<div
  className="tipseenBasicContent"
  data-testid="tipseen-content"
>
  <div
    className="buttons"
  />
</div>
`;

exports[`Snapshot tests > Tipseen tests > renders correctly with dark close button theme 1`] = `
<DocumentFragment>
  <div
    class="monday-storybook-tipseen_container"
  />
</DocumentFragment>
`;

exports[`Snapshot tests > Tipseen tests > renders correctly with floating variation 1`] = `null`;

exports[`Snapshot tests > Tipseen tests > renders correctly without close 1`] = `
<DocumentFragment>
  <div
    class="monday-storybook-tipseen_container"
  />
</DocumentFragment>
`;

exports[`Snapshot tests > Tipseen tests > renders correctly without props 1`] = `
<div
  className="monday-storybook-tipseen_container"
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
/>
`;

exports[`Snapshot tests > TipseenTitle > should render correctly with given text 1`] = `
<div
  aria-level={3}
  className="typography inherit start multiLineEllipsis text text1Bold title"
  data-testid="tipseen-title"
  role="heading"
  style={
    {
      "--text-clamp-lines": "2",
    }
  }
>
  I'm a title
</div>
`;

exports[`Snapshot tests > TipseenTitle > should render null without text 1`] = `null`;
